packages/contracts-bedrock/contracts/dispute/AttestationDisputeGame.sol

```diff
@@ -0,0 +1,287 @@
+        if (challenges[recovered]) {
+            revert AlreadyChallenged();
+        }
+
+        // Mark that the authorized signer has issued a positive attestation for the `rootClaim`.
+        challenges[recovered] = true;
+
+        // Increment the number of positive attestations that have been issued for the `rootClaim`.
+        attestationSubmitters.push(msg.sender);
+
+        // If the provided signature breaches the signature threshold, resolve the game.
+        if (attestationSubmitters.length == frozenSignatureThreshold) {
+            resolve();
+        }
+    }
+
+    /**
+     * @notice Returns an Ethereum Signed Typed Data hash, as defined in EIP-712, for the
+     *         `Dispute` struct. This hash is signed by members of the `attestorSet` to
+     *         issue a positive attestation for the `rootClaim`.
+     * @return _typedDataHash The EIP-712 hash of the `Dispute` struct.
+     */
+    function getTypedDataHash() public view returns (Hash _typedDataHash) {
+        // Copy the `DISPUTE_TYPE_HASH` onto the stack.
+        Hash disputeTypeHash = DISPUTE_TYPE_HASH;
+        // Grab the root claim of the `AttestationDisputeGame`.
+        Claim _rootClaim = rootClaim();
+        // Grab the L2 block number that the `rootClaim` commits to.
+        uint256 _l2BlockNumber = l2BlockNumber();
+
+        // Hash the `Dispute` struct.
+        Hash disputeStructHash;
+        assembly {
+            // Grab the location of some free memory.
+            let ptr := mload(0x40)
+
+            // Store the `DISPUTE_TYPE_HASH`
+            mstore(ptr, disputeTypeHash)
+            // Store the `rootClaim` of the `AttestationDisputeGame`.
+            mstore(add(ptr, 0x20), _rootClaim)
+            // Store the L2 block number that the `rootClaim` commits to.
+            mstore(add(ptr, 0x40), _l2BlockNumber)
+
+            disputeStructHash := keccak256(ptr, 0x60)
+        }
+
+        _typedDataHash = Hash.wrap(_hashTypedData(Hash.unwrap(disputeStructHash)));
+    }
+
+    /**
+     * @notice Initializes the contract.
+     */
+    function initialize() external initializer {
+        createdAt = Timestamp.wrap(uint64(block.timestamp));
+        frozenAttestorSet = systemConfig.attestorSet();
+        frozenSignatureThreshold = systemConfig.attestationThreshold();
+    }
+
+    /**
+     * @notice Returns the semantic version of the DisputeGame contract.
+     * @dev Current version: 0.0.1
+     * @return The semantic version of the DisputeGame contract.
+     */
+    function version() external pure override returns (string memory) {
+        assembly {
+            // Store the pointer to the string
+            mstore(returndatasize(), 0x20)
+            // Store the version ("0.0.1")
+            // len |   "0.0.1"
+            // 0x05|302E302E31
+            mstore(0x25, 0x05302E302E31)
+            // Return the semantic version of the contract
+            return(returndatasize(), 0x60)
```

Comment:
wen push0

Reply:
Not opposed to updating the solc version to 0.8.20 for the new set of contracts, should certainly not do that for the other contracts tho until we are deliberate about it

Reply:
Yeah I lean towards keeping 0.8.15 until we've done a mass upgrade. Big push0 fanboy, but introducing another compiler version into our compilation pipeline would be a bit annoying. wdyt?

Reply:
(Another note- this function is extra asf. Leaning towards just returning an immutable string lmfao.)